EHN Normalizing the last layer in all the models

brunaafl opened the pull request:
Changed last layers' names to final_layer, cf #457.

Fix TIDNet HybridNet's and inheritance (open issue).

sliwy commented:
@brunaafl You may want to take a look at #513 as there may be some conflicts and try to coordinate/see what are the common parts :)

brunaafl commented:
> @brunaafl You may want to take a look at #513 as there may be some conflicts and try to coordinate/see what are the common parts :)

Thanks! I will!

brunaafl commented:
TODO: Add a new way to handle return features

During the renaming, I had to remove the possibility of returning the output of the feature extractor when there was this option (on the SleepStagerEldele2021, SleepStagerBlanco2020, SleepStagerChambon2018, and DeepSleepNet models). However, this is required for some tutorials, so now they are broken.

PierreGtch commented:
@brunaafl on it

PierreGtch commented:
can you update the whats_new?

brunaafl commented:
> can you update the whats_new?

Yes! I'll do that

sliwy commented:
thx for the PR @brunaafl 

We discussed with @PierreGtch more in details how it should be done to support the transfer learning approach and be easy to use. We identified few cases in which we may act in slightly modified manner, so the PR will require some changes. We would be really happy if you find some time and try to address the changes:

1. The case in which we have `nn.Module` -> we would like to have the last layer providing the tensor with on of the dimensions equal to `n_outputs` to be called `final_layer` and all the modules after should be also incorporated into the `nn.Sequential` with it. Sometimes it requires splitting `nn.Sequential` into two parts or merging `nn.Linear` with some reshaping. Reshaping should be included in the `final_layer` because if we replace `final_layer` with `nn.Identity` then it will break the code because of mismatch of dimensions.
2. If the model is `nn.Sequential` we would like to have the last layer providing the tensor with on of the dimensions equal to `n_outputs` to be called `final_layer` and all the modules after incorporated into the same module.
3. `LogSoftmax` -> we would like to not touch `LogsSoftmax` as it is going to be renamed in another PR and it's going to solve a different problem. Let's assume that `LogSoftmax` is not changing the shape of the tensor and it should not be included in the `final_layer`.

@brunaafl would you have time to tackle that? if you have any doubts I would be happy to help! :smile:

brunaafl commented:
> thx for the PR @brunaafl
> 
> We discussed with @PierreGtch more in details how it should be done to support the transfer learning approach and be easy to use. We identified few cases in which we may act in slightly modified manner, so the PR will require some changes. We would be really happy if you find some time and try to address the changes:
> 
> 1. The case in which we have `nn.Module` -> we would like to have the last layer providing the tensor with on of the dimensions equal to `n_outputs` to be called `final_layer` and all the modules after should be also incorporated into the `nn.Sequential` with it. Sometimes it requires splitting `nn.Sequential` into two parts or merging `nn.Linear` with some reshaping. Reshaping should be included in the `final_layer` because if we replace `final_layer` with `nn.Identity` then it will break the code because of mismatch of dimensions.
> 2. If the model is `nn.Sequential` we would like to have the last layer providing the tensor with on of the dimensions equal to `n_outputs` to be called `final_layer` and all the modules after incorporated into the same module.
> 3. `LogSoftmax` -> we would like to not touch `LogsSoftmax` as it is going to be renamed in another PR and it's going to solve a different problem. Let's assume that `LogSoftmax` is not changing the shape of the tensor and it should not be included in the `final_layer`.
> 
> @brunaafl would you have time to tackle that? if you have any doubts I would be happy to help! 😄

Hi! Yes, I can address those changes, although not today, just tomorrow and after. I'm happy to help!

brunaafl commented:
> > thx for the PR @brunaafl
> > We discussed with @PierreGtch more in details how it should be done to support the transfer learning approach and be easy to use. We identified few cases in which we may act in slightly modified manner, so the PR will require some changes. We would be really happy if you find some time and try to address the changes:
> > 
> > 1. The case in which we have `nn.Module` -> we would like to have the last layer providing the tensor with on of the dimensions equal to `n_outputs` to be called `final_layer` and all the modules after should be also incorporated into the `nn.Sequential` with it. Sometimes it requires splitting `nn.Sequential` into two parts or merging `nn.Linear` with some reshaping. Reshaping should be included in the `final_layer` because if we replace `final_layer` with `nn.Identity` then it will break the code because of mismatch of dimensions.
> > 2. If the model is `nn.Sequential` we would like to have the last layer providing the tensor with on of the dimensions equal to `n_outputs` to be called `final_layer` and all the modules after incorporated into the same module.
> > 3. `LogSoftmax` -> we would like to not touch `LogsSoftmax` as it is going to be renamed in another PR and it's going to solve a different problem. Let's assume that `LogSoftmax` is not changing the shape of the tensor and it should not be included in the `final_layer`.
> > 
> > @brunaafl would you have time to tackle that? if you have any doubts I would be happy to help! 😄
> 

HI! I was wondering here, just to know if I really understood. So, you want the first layer that changes the shape of the tensor to one with one of the dimensions equal to n_outputs to be called `final_layer`, and all the other modules after this one incorporated into the same sequential with it. Is that right? And the `LogSoftmax` should not be included in this.

My doubt is in the case where `LogSoftmax` comes in between a conv that changes the shape of one of the dimensions to n_outputs and another layer (such as one to squeeze the final output), such as in the case of Deep4Net and EEGNet. Should I just leave `LogSoftmax` inside the `final_layer` module, since it will be changed in a next pr?

PierreGtch commented:
@brunaafl sorry for the confusion, there was some miscommunication at multiple levels. 

So everything after and including the classification layer (sometimes it is a`nn.Linear`, sometimes `nn.Conv2d`..) must go into this new `final_layer`.
Typically you will find a softmax, a reshape, a transpose, or a squeeze, etc. They should all go inside `final_layer`. For example, `Deep4net` will change from:
```python
        self.add_module(
            "conv_classifier",
            nn.Conv2d(
                self.n_filters_4,
                self.n_outputs,
                (self.final_conv_length, 1),
                bias=True,
             ),
        )
        self.add_module("softmax", nn.LogSoftmax(dim=1))
        self.add_module("squeeze", Expression(squeeze_final_output))
```
TO:
```python
        self.add_module("final_layer", nn.Sequential())
        self.final_layer.add_module(
            "conv_classifier",
            nn.Conv2d(
                self.n_filters_4,
                self.n_outputs,
                (self.final_conv_length, 1),
                bias=True,
             ),
        )
        self.final_layer.add_module("softmax", nn.LogSoftmax(dim=1))
        self.final_layer.add_module("squeeze", Expression(squeeze_final_output))
```
If it's still not clear, please tell me

brunaafl commented:
> @brunaafl sorry for the confusion, there was some miscommunication at multiple levels.
> 
> So everything after and including the classification layer (sometimes it is a`nn.Linear`, sometimes `nn.Conv2d`..) must go into this new `final_layer`. Typically you will find a softmax, a reshape, a transpose, or a squeeze, etc. They should all go inside `final_layer`. For example, `Deep4net` will change from:
> 
> ```python
>         self.add_module(
>             "conv_classifier",
>             nn.Conv2d(
>                 self.n_filters_4,
>                 self.n_outputs,
>                 (self.final_conv_length, 1),
>                 bias=True,
>              ),
>         )
>         self.add_module("softmax", nn.LogSoftmax(dim=1))
>         self.add_module("squeeze", Expression(squeeze_final_output))
> ```
> 
> TO:
> 
> ```python
>         self.add_module("final_layer", nn.Sequential())
>         self.final_layer.add_module(
>             "conv_classifier",
>             nn.Conv2d(
>                 self.n_filters_4,
>                 self.n_outputs,
>                 (self.final_conv_length, 1),
>                 bias=True,
>              ),
>         )
>         self.final_layer.add_module("softmax", nn.LogSoftmax(dim=1))
>         self.final_layer.add_module("squeeze", Expression(squeeze_final_output))
> ```
> 
> If it's still not clear, please tell me

Okay, it's clear now, thanks!

PierreGtch commented:
Good point @robintibor, I opened an issue: #532

brunaafl commented:
Hi! 
So I changed the last layers' names just on the Sequential models, for now.
I also added a new function to the base file, return_new_keys, where I just changed the old state dictionary to the updated one and return it. Inside each model's class, I put a load_state_dict that just calls this function and update the dict. But I don't know if this is the best way to do this.

bruAristimunha commented on braindecode/models/base.py:
```diff
@@ -197,6 +199,18 @@ def get_output_shape(self) -> Tuple[int]:
                 raise exc
 
+    def return_new_keys(self, state_dict, keys_to_change):  # personalised keys_to_change
+        """Wrapper to allow for loading of a state_dict from a model before CombinedConv was
+         implemented"""
```
doc

bruAristimunha commented on braindecode/models/base.py:
```diff
@@ -197,6 +199,18 @@ def get_output_shape(self) -> Tuple[int]:
                     raise ValueError(msg) from exc
                 raise exc
 
+    def return_new_keys(self, state_dict, keys_to_change):  # personalised keys_to_change
```
add test

sliwy replied:
I would also remove the comment in the end of line -> if there is sth user should know put it in the docstring

sliwy replied:
Let's change the method name to sth that contains final_layer, for example -> `map_keys_final_layer`

bruAristimunha commented on braindecode/models/deep4.py:
```diff
@@ -270,17 +270,18 @@ def add_conv_pool_block(
-        self.add_module("softmax", nn.LogSoftmax(dim=1))
-        self.add_module("final_layer", Expression(squeeze_final_output))
+
+        # The conv_classifier will be the final_layer and the other ones will be incorporated
+        # The conv_classifier will be the final_layer and the other ones will be incorporated
```
improve the doc

bruAristimunha commented on braindecode/models/eeginception.py:
```diff
@@ -244,14 +244,22 @@ def __init__(
+        # Moved flatten to another layer
+        self.add_module("flat", nn.Flatten())
+
+        # The conv_classifier will be the final_layer and the other ones will be incorporated
```
improve the doc here

bruAristimunha commented on braindecode/models/eeginception_erp.py:
```diff
@@ -240,14 +240,22 @@ def __init__(
+
+        module.add_module("softmax", nn.Softmax(1))
+
+        # The conv_classifier will be the final_layer and the other ones will be incorporated
```
same

bruAristimunha commented on braindecode/models/eegnet.py:
```diff
@@ -200,10 +222,17 @@ def __init__(
         _glorot_weight_zero_bias(self)
 
+    def load_state_dict(self, state_dict, *args, **kwargs):
+        """Wrapper to allow for loading of a state_dict from a model before CombinedConv was
+         implemented"""
```
doc

bruAristimunha commented on braindecode/models/atcnet.py:
good!

bruAristimunha commented on braindecode/models/eegnet.py:
```diff
@@ -345,6 +378,25 @@ def __init__(
             n_out_time = output_shape[3]
             self.final_conv_length = n_out_time
 
+        # The conv_classifier will be the final_layer and the other ones will be incorporated
+        module = nn.Sequential()
+
+        module.add_module("conv_classifier",
+                          nn.Conv2d(n_filters_3, self.n_outputs,
+                                    (n_out_virtual_chans, self.final_conv_length), bias=True, ))
+
+        module.add_module("softmax", nn.LogSoftmax(dim=1))
+
+        # Transpose back to the the logic of braindecode,
+        # so time in third dimension (axis=2)
+        module.add_module("permute_2", Rearrange("batch x y z -> batch x z y"),)
+
+        module.add_module("squeeze", Expression(squeeze_final_output))
+
+        # The conv_classifier will be the final_layer and the other ones will be incorporated
+        self.add_module("final_layer", module)
+
+        """
```
Looks strange in the github

bruAristimunha commented on braindecode/models/eegnet.py:
```diff
@@ -360,9 +412,16 @@ def __init__(
         )
-        self.add_module("squeeze", Expression(squeeze_final_output))
+
+        self.add_module("final_layer", Expression(squeeze_final_output))"""
```
why is commented?

bruAristimunha commented on braindecode/models/eegresnet.py:
```diff
@@ -188,11 +188,29 @@ def __init__(
+        """
         self.add_module('conv_classifier',
-                        nn.Conv2d(n_cur_filters, self.n_outputs,
-                                  (1, 1), bias=True))
+                nn.Conv2d(n_cur_filters, self.n_outputs,
+                          (1, 1), bias=True))
         self.add_module('softmax', nn.LogSoftmax(dim=1))
-        self.add_module('squeeze', Expression(squeeze_final_output))
+
+        self.add_module('final_layer', Expression(squeeze_final_output))
+
+        """
```
same here

bruAristimunha commented on test/unit_tests/models/test_models.py:
```diff
@@ -456,7 +456,7 @@ def test_sleep_stager_return_feats():
 
     out = model(X)
-    assert out.shape == (10, model.len_last_layer)
+    assert out.shape == (10, model.len_last_layer)"""
```
Insert a comment why is commented

bruAristimunha commented on test/unit_tests/models/test_models.py:
```diff
@@ -492,7 +492,7 @@ def test_eldele_2021(sfreq, n_classes, input_size_s, d_model):
 
 
-def test_eldele_2021_feats():
+"""def test_eldele_2021_feats():
```
same here

bruAristimunha reviewed (changes requested):
Small details and make more test for the new function. Look's nice! Almost merged!

bruAristimunha commented:
Don't forget to resolve conflicts when you finish.
![image](https://github.com/braindecode/braindecode/assets/42702466/74a06b0d-7ee1-4215-a3b4-ce5d30538ccd)

sliwy commented on braindecode/models/base.py:
```diff
@@ -197,6 +199,18 @@ def get_output_shape(self) -> Tuple[int]:
+                k = f"final_layer.{k}"
+            new_state_dict[k] = v
+
+        return new_state_dict  # not module to return state dict
```
I don't understand this comment, I'd remove it anyway

sliwy commented on braindecode/models/deep4.py:
```diff
@@ -308,24 +319,27 @@ def add_conv_pool_block(
-            new_state_dict[k] = v
+            state_dict_time_spat[k] = v
+
+        # I just added this to dan's function
```
I would remove this kind of comments, only code and some important info for the code that could be hard to understand. If you want to let us know something just leave a comment in PR :smiley: 

```suggestion
```

sliwy commented on braindecode/models/deep4.py:
```diff
@@ -270,17 +278,20 @@ def add_conv_pool_block(
-        self.add_module("softmax", nn.LogSoftmax(dim=1))
-        self.add_module("squeeze", Expression(squeeze_final_output))
+
+        # The conv_classifier will be the final_layer and the other ones will be incorporated
+        module = nn.Sequential()
+
+        module.add_module("conv_classifier",
+                          nn.Conv2d(n_filters_4, self.n_outputs,
+                                    (self.final_conv_length, 1), bias=True, ))
+
+        module.add_module("softmax", nn.LogSoftmax(dim=1))
+
+        module.add_module("squeeze", Expression(squeeze_final_output))
+
+        # The conv_classifier will be the final_layer and the other ones will be incorporated
+        self.add_module("final_layer", module)
```
For this part i would use the code snippet provided by @PierreGtch https://github.com/braindecode/braindecode/pull/520#issuecomment-1713563121

sliwy commented on braindecode/models/deep4.py:
```diff
@@ -308,24 +319,27 @@ def add_conv_pool_block(
 
-        init.xavier_uniform_(self.conv_classifier.weight, gain=1)
-        init.constant_(self.conv_classifier.bias, 0)
+        init.xavier_uniform_(self.final_layer[0].weight, gain=1)
```
```suggestion
        init.xavier_uniform_(self.final_layer.conv_classifier.weight, gain=1)
```
Let's also adjust here to use explicit layer names

sliwy commented on braindecode/models/deep4.py:
```diff
@@ -308,24 +319,27 @@ def add_conv_pool_block(
-        init.xavier_uniform_(self.conv_classifier.weight, gain=1)
-        init.constant_(self.conv_classifier.bias, 0)
+        init.xavier_uniform_(self.final_layer[0].weight, gain=1)
+        init.constant_(self.final_layer[0].bias, 0)
```
```suggestion
        init.constant_(self.final_layer.conv_classifier.bias, 0)
```

sliwy commented on braindecode/models/eeginception.py:
```diff
@@ -244,17 +249,33 @@ def __init__(
-            ),
-            nn.Softmax(1)
-        ))
+        # Moved flatten to another layer
```
same here with the comment, I'd remove

sliwy commented on braindecode/models/eeginception.py:
```diff
@@ -244,17 +249,33 @@ def __init__(
                 self.n_times // prod(self.pooling_sizes))
         n_channels_last_layer = self.n_filters * len(self.scales_samples) // 4
 
-        self.add_module("classification", nn.Sequential(
```
I would just rename `"classification"` to `"final_layer"` and keep the flatten separated (added just before, same as you did)

sliwy commented on braindecode/models/eeginception_erp.py:
```diff
@@ -240,17 +245,33 @@ def __init__(
                 self.n_times // prod(self.pooling_sizes))
         n_channels_last_layer = self.n_filters * len(self.scales_samples) // 4
 
-        self.add_module("classification", nn.Sequential(
```
I would just rename "classification" to "final_layer" and keep the flatten separated (added just before, same as you did)

sliwy commented on braindecode/models/eeginception_mi.py:
```diff
@@ -171,7 +171,7 @@ def __init__(
         )
 
-        self.softmax = nn.LogSoftmax(dim=1)
+        self.final_layer = nn.LogSoftmax(dim=1)
```
`self.fc` should be also in the `final_layer`

sliwy commented on braindecode/models/eegitnet.py:
```diff
@@ -197,10 +202,21 @@ def __init__(
+                        nn.Linear(int(int(self.n_times / 4) / 4) * 28, self.n_outputs))
+
+        # Softmax will be changed, so I took it put of final_layer
+        self.add_module("sfmx", nn.Softmax(dim=1))
```
softmax should be inside `final_layer`

sliwy commented on braindecode/models/eegnet.py:
```diff
@@ -186,24 +193,33 @@ def __init__(
-        self.add_module("softmax", nn.LogSoftmax(dim=1))
-        # Transpose back to the the logic of braindecode,
+        # The conv_classifier will be the final_layer and the other ones will be incorporated
+        module = nn.Sequential()
```
I would go with the code similar to one of @PierreGtch :
```
        self.final_layer = nn.Sequential()
        self.final_layer.add_module("conv_classifier",
                          nn.Conv2d(self.F2, self.n_outputs,
                                    (n_out_virtual_chans, self.final_conv_length), bias=True, ))
        self.final_layer.add_module("softmax", nn.LogSoftmax(dim=1))
        self.final_layer.add_module("permute_back", Rearrange("batch x y z -> batch x z y"), )

        self.final_layer.add_module("squeeze", Expression(squeeze_final_output))
```

sliwy commented on braindecode/models/eegnet.py:
```diff
@@ -345,24 +368,33 @@ def __init__(
-        )
-        self.add_module("softmax", nn.LogSoftmax(dim=1))
+        # The conv_classifier will be the final_layer and the other ones will be incorporated
+        module = nn.Sequential()
```
same here with the way how we add the `self.final_layer`

sliwy commented on braindecode/models/eegresnet.py:
```diff
@@ -188,18 +193,34 @@ def __init__(
-        self.add_module('squeeze', Expression(squeeze_final_output))
+
+        # The conv_classifier will be the final_layer and the other ones will be incorporated
+        module = nn.Sequential()
```
same with `self.final_layer`

sliwy commented on braindecode/models/hybrid.py:
```diff
@@ -109,7 +110,7 @@ def forward(self, x):
 
         merged_out = torch.cat((deep_out, shallow_out), dim=1)
-        linear_out = self.final_conv(merged_out)
+        linear_out = self.final_layer(merged_out)
         softmaxed = nn.LogSoftmax(dim=1)(linear_out)
         squeezed = softmaxed.squeeze(3)
```
These should be included in `final_layer`. Squeezing may be replaced with `Expression`?

brunaafl replied:
Done. I left the squeeze there, if you all decide to just replace with expression let me know

sliwy commented on braindecode/models/shallow_fbcsp.py:
```diff
@@ -170,17 +178,20 @@ def __init__(
-        self.add_module("squeeze", Expression(squeeze_final_output))
+
+        # The conv_classifier will be the final_layer and the other ones will be incorporated
+        module = nn.Sequential()
```
same here with `final_layer`

sliwy commented on braindecode/models/shallow_fbcsp.py:
```diff
@@ -194,21 +205,23 @@ def __init__(
             init.constant_(self.bnorm.bias, 0)
-        init.xavier_uniform_(self.conv_classifier.weight, gain=1)
-        init.constant_(self.conv_classifier.bias, 0)
+        init.xavier_uniform_(self.final_layer[0].weight, gain=1)
```
and here I would also use directly the name of the module - `conv_classifier`, just in case someone adds one more layer in the beginning  of `final_layer` then the code won't crash

sliwy commented on braindecode/models/tcn.py:
```diff
@@ -90,7 +90,8 @@ def __init__(
         self.fc = nn.Linear(in_features=n_filters, out_features=self.n_outputs)
         if add_log_softmax:
             self.log_softmax = nn.LogSoftmax(dim=1)
-        self.squeeze = Expression(squeeze_final_output)
+
+        self.final_layer = Expression(squeeze_final_output)
```
I think we should add `self.fc` `log_softmax`, and `squeeze` to `final_layer

sliwy commented on examples/model_building/plot_hyperparameter_tuning_with_scikit-learn.py:
```diff
@@ -348,7 +348,7 @@
 best_run = search_results[search_results['rank_test_score'] == 1].squeeze()
 print(
-    f"Best hyperparameters were {best_run['params']} which gave a validation "
+    f"The best hyperparameters gave a validation "
```
why do we modify this? if it's intentional we you should remove the `f-string` in front

sliwy reviewed (changes requested):
Hey @brunaafl, thanks for improving on the PR :smile: , I marked few style changes and one or two spots where we have to update the `final_layer` to include more layers, looking forward your changes :pray:

sliwy commented on braindecode/models/atcnet.py:
```diff
@@ -207,23 +207,25 @@ def __init__(
 
         if self.concat:
-            self.max_norm_linears = nn.ModuleList([
+            self.final_layer = nn.ModuleList([
```
@PierreGtch @brunaafl this one is quite tricky because it has different ways of processing depending on `self.concat` and so on. To really fulfill all the requirements we may need to create a custom `nn.Module` that would serve as the final layer`. I may think of sth tomorrow evening

PierreGtch commented:
Great work @brunaafl ! I agree with the comments of @bruAristimunha and @sliwy and have one of my own:
Instead of having just `self.keys_to_change` as a list, I would replace it by a dict mapping old keys to new keys. This would be a bit more generic and allow to handle the case of `Deep4Net` that also overwrites some keys. 
Then I would overwrite `load_state_dict` directly in `EEGModuleMixin` (and eventually remove ` return_new_keys`) so that you don't have duplicate code.

brunaafl reviewed (commented):
done

bruAristimunha commented on braindecode/models/eegconformer.py:
```diff
@@ -389,9 +428,10 @@ def __init__(self, final_fc_length, n_classes,
             return out, x
         else:
-            return self.classification(out)
+            self.final_layer.add_module('classification', nn.LogSoftmax(dim=1))
```
```suggestion

```

bruAristimunha replied:
```suggestion
            self.final_layer.add_module('classification', self.classification)
```

bruAristimunha commented on braindecode/models/eegconformer.py:
```diff
@@ -380,6 +387,38 @@ def __init__(self, final_fc_length, n_classes,
+
+        Parameters
+        ----------
+        final_fc_length : int
+            Length of the final fully connected layer.
+        n_classes : int
+            Number of classes for classification.
+        drop_prob_1 : float
+            Dropout probability for the first dropout layer.
+        drop_prob_2 : float
+            Dropout probability for the second dropout layer.
+        out_channels : int
+            Number of output channels for the first linear layer.
```
```suggestion
        n_classes : int
            Number of classes for classification.
```

bruAristimunha commented on braindecode/models/eegconformer.py:
```diff
@@ -380,6 +387,38 @@ def __init__(self, final_fc_length, n_classes,
+            Number of output channels for the first linear layer.
+        hidden_channels : int
+            Number of output channels for the second linear layer.
+        return_features : bool
+            Whether to return input features.
```
```suggestion
        return_features : bool
            Whether to return input features.
        add_log_softmax : bool
            Adding LogSoftmax or not.
```

bruAristimunha commented on braindecode/models/eeginception.py:
```diff
@@ -246,17 +250,33 @@ def __init__(
 
         _glorot_weight_zero_bias(self)
 
+    def load_state_dict(self, state_dict, *args, **kwargs):
+        """Wrapper to allow for loading of a state_dict from a model before CombinedConv was
+         implemented"""
+
+        new_state_dict = super().return_new_keys(state_dict, self.keys_to_change)
+        return super().load_state_dict(new_state_dict, *args, **kwargs)
+
```
```suggestion
```

bruAristimunha commented on braindecode/models/eegnet.py:
```diff
@@ -186,25 +197,33 @@ def __init__(
 
         _glorot_weight_zero_bias(self)
 
+    """def load_state_dict(self, state_dict, *args, **kwargs):
+         # Wrapper to allow for loading of a state_dict from a model before CombinedConv was
+         # implemented and the las layers' names were normalized
+
+        new_state_dict = super().return_new_keys(state_dict, self.keys_to_change)
+        return super().load_state_dict(new_state_dict, *args, **kwargs)"""
+
```
```suggestion

```

bruAristimunha commented on braindecode/models/eegnet.py:
```diff
@@ -100,6 +100,17 @@ def __init__(
+        # For the load_state_dict
+        # When padronize all layers,
+        # add the old's parameters here
+        """self.keys_to_change = [
+            "conv_classifier.weight",
+            "conv_classifier.bias"
+        ]"""
```
```suggestion

```

bruAristimunha commented:
All good for me @brunaafl. Amazing working, normalizing the last layer in all the models. @sliwy, when you approve, it will be merged =)

PierreGtch commented on braindecode/models/deepsleepnet.py:
```diff
@@ -222,6 +222,8 @@ def __init__(
         self.return_feats = return_feats
+
+        # TODO: Add new way to handle return_features == True
         if not return_feats:
             self.final_layer = nn.Linear(1024, self.n_outputs)
```
```suggestion
        if not return_feats:
            self.final_layer = nn.Linear(1024, self.n_outputs)
        else:
            self.final_layer = nn.Identity()
```

PierreGtch commented on braindecode/models/base.py:
```diff
@@ -215,6 +217,18 @@ def get_output_shape(self) -> Tuple[int]:
                     raise ValueError(msg) from exc
                 raise exc
 
+    def load_state_dict(self, state_dict, *args, **kwargs):
+
+        mapping = self.mapping
```
```suggestion
    mapping = None
    
    def load_state_dict(self, state_dict, *args, **kwargs):

        mapping = self.mapping if self.mapping else {}
```
So that we don’t have to add this mapping attribute to the models that don’t need if

PierreGtch commented on braindecode/models/eegconformer.py:
```diff
@@ -149,15 +154,18 @@ def __init__(
             att_drop=att_drop_prob)
 
         self.classification_head = _ClassificationHead(
```
```suggestion
        self.fc = _FullyConnected(
```
The name `_ClassificationHead` does not make any more sense

PierreGtch commented on braindecode/models/eegconformer.py:
```diff
@@ -149,15 +154,18 @@ def __init__(
         x = torch.unsqueeze(x, dim=1)  # add one extra dimension
         x = self.patch_embedding(x)
         x = self.transformer(x)
         x = self.classification_head(x)
```
```suggestion
        x = fc(x)
```

PierreGtch commented on braindecode/models/eegconformer.py:
```diff
@@ -346,10 +354,9 @@ def __init__(self, att_depth, emb_size, att_heads, att_drop):
 
 
 class _ClassificationHead(nn.Module):
```
```suggestion
class _FullyConnected(nn.Module):
```

PierreGtch commented on braindecode/models/eegconformer.py:
```diff
@@ -346,10 +354,9 @@ def __init__(self, att_depth, emb_size, att_heads, att_drop):
-                 hidden_channels=32, return_features=False,
-                 add_log_softmax=True):
+                 hidden_channels=32):
         """"Classification head for the transformer encoder.
```
```suggestion
        "”””Fully-connected layer for the transformer encoder.
```

PierreGtch commented on braindecode/models/hybrid.py:
```diff
@@ -101,9 +105,19 @@ def __init__(self, n_chans=None, n_outputs=None, n_times=None,
+
+        self.final_layer = module
+        """self.final_layer = nn.Conv2d(
             100, self.n_outputs, kernel_size=(1, 1), stride=1
-        )
+        )"""
```
Can you remove this?

PierreGtch reviewed (changes requested):
Hi @brunaafl, great work! I left a few suggestions, otherwise all good for me to merge.

In another PR, we will need to add at least one test per model to check if they can be initialised and produce the correct output shape.

sliwy commented on braindecode/models/eegconformer.py:
```diff
@@ -389,9 +422,10 @@ def __init__(self, final_fc_length, n_classes,
             return out, x
         else:
-            return self.classification(out)
+            self.final_layer.add_module('classification', self.classification)
```
If we introduce changes from above adding layer in forward will not be needed
```suggestion
```

sliwy commented on braindecode/models/eegconformer.py:
```diff
@@ -389,9 +422,10 @@ def __init__(self, final_fc_length, n_classes,
             self.classification = nn.Identity()
 
```
```suggestion
if not self.return_features:
    self.final_layer.add_module("classification", classification)
```

sliwy commented on braindecode/models/eegconformer.py:
```diff
@@ -389,9 +422,10 @@ def __init__(self, final_fc_length, n_classes,
             self.classification = nn.Identity()
```
here, assign `softmax`/`identity` to just `classification` variable and not `self.classification`, then when looking at the model layers `classification` will apear only once not twice

sliwy commented on braindecode/models/hybrid.py:
```diff
@@ -101,9 +105,16 @@ def __init__(self, n_chans=None, n_outputs=None, n_times=None,
-            100, self.n_outputs, kernel_size=(1, 1), stride=1
-        )
+
+        module = nn.Sequential(
+            nn.Conv2d(
+                100,
+                self.n_outputs,
+                kernel_size=(1, 1),
+                stride=1),
+            nn.LogSoftmax(dim=1) if self.add_log_softmax else nn.Identity())
+
+        self.final_layer = module
```
```suggestion
        self.final_layer = nn.Sequential(
            nn.Conv2d(
                100,
                self.n_outputs,
                kernel_size=(1, 1),
                stride=1),
            nn.LogSoftmax(dim=1) if self.add_log_softmax else nn.Identity())

```

sliwy commented on braindecode/models/shallow_fbcsp.py:
```diff
@@ -197,21 +210,5 @@ def __init__(
-                k = f"conv_time_spat.{k}"
-            new_state_dict[k] = v
-        return super().load_state_dict(new_state_dict, *args, **kwargs)
+        init.xavier_uniform_(self.final_layer[0].weight, gain=1)
+        init.constant_(self.final_layer[0].bias, 0)
```
```suggestion
        init.xavier_uniform_(self.final_layer.conv_classifier.weight, gain=1)
        init.constant_(self.final_layer.conv_classifier.bias, 0)
```

sliwy commented on braindecode/models/tcn.py:
```diff
@@ -71,6 +71,15 @@ def __init__(
+            "fc.weight": "final_layer.fc.weight",
+            "fc.bias": "final_layer.fc.bias"
+        }
+        self.keys_to_change = [
+            'fc.weight',
+            'fc.bias'
+        ]
+
```
```suggestion
```

sliwy commented on braindecode/models/eegitnet.py:
```diff
@@ -147,6 +147,10 @@ def __init__(
             sfreq=sfreq,
             add_log_softmax=add_log_softmax,
         )
+        self.keys_to_change = {
```
```suggestion
        self.mapping = {
```

sliwy reviewed (changes requested):
Thanks @brunaafl for new changes! :)

I added few more comments. Most important, in `eegconformer.py` there are 4 `"` in two docstrings, so we have to fix that before merge (I have added suggestions). Then, few details with naming attributes or removing old attributes.

@bruAristimunha do you have any idea why tests didn't run on this PR after last commit?

bruAristimunha commented:
Thank you @brunaafl, @sliwy and @PierreGtch
